migrator/zenodo_rdm_migrator/extract/kafka.py

```diff
@@ -7,33 +7,291 @@
+        """Yields operations/statements."""
+        consumer = self._ops_consumer
+        for op_msg in consumer:
+            self._dump_msg(self.ops_topic, op_msg)
+
+            if not op_msg.value:
+                self.logger.debug(f"No message value for op {op_msg}")
+                consumer.commit()
+                continue
+            tx_id = op_msg.value["source"]["txId"]
+            if tx_id <= self.last_tx:
+                consumer.commit()
+                continue
+            consumer.commit()
+            yield (tx_id, dict(key=op_msg.key, **op_msg.value))
+
+    def _yield_completed_tx(self, min_batch=None):
+        """Yields completed transactions."""
+        completed_tx_batch = []
+        for tx_state in self.tx_registry.values():
+            if not tx_state.complete:
+                # We stop at the first non-completed transaction
+                break
+            completed_tx_batch.append(tx_state)
+
+            # if we reached the minimum batch size, we stop
+            if min_batch and len(completed_tx_batch) == min_batch:
+                break
+
+        # If we didn't make a big enough batch we return
+        if min_batch and len(completed_tx_batch) < min_batch:
+            return
+
+        for tx in completed_tx_batch:
+            del self.tx_registry[tx.id]
+            # Keep track of the last yielded transaction ID
+            self._last_yielded_tx = tx.id
+            yield Tx(id=tx.id, operations=list(tx.ops))
 
     def run(self):
-        """Yield one element at a time."""
-        consumer = KafkaConsumer(*self.topics, **self.kafka_cfg)
-        yield from consumer
+        """Return a blocking generator yielding completed transactions."""
+        # We're using an (always) SortedDict, since we want to yield transactions in
+        # their correct order.
+        try:
+            while True:
+                # First we consume all the available transactions information
+                for tx_id, tx_info in self.iter_tx_info():
+                    if tx_id in self.tx_registry:
+                        self.tx_registry[tx_id].info = tx_info
+                    else:
+                        self.tx_registry[tx_id] = TxState(tx_id, tx_info)
+
+                # We now consume operations and build on a (pending) transaction state.
+                for tx_id, op in self.iter_ops():
+                    tx_state = self.tx_registry.setdefault(tx_id, TxState(tx_id))
+                    tx_state.append(op)
+                    if tx_state.complete:
+                        self.logger.info(f"Completed transaction {tx_state.id}")
+
+                # If no new transactions, we don't need to sleep since consumers
+                # have a timeout/sleep already via "consumer_timeout_ms".
+                yield from self._yield_completed_tx(min_batch=self.tx_buffer)
+        # Normally this extract would run forever, so we need some mechanism to stop
+        except KafkaExtractEnd:
```

Comment:
understanding check: when would we hit this? at some point we would "switch off" old Zenodo and then this would happen. That's it right?

Reply:
Exactly, when we shut down legacy Zenodo, we'll just stop seeing new transactions coming in. At that point we hit Ctr-C (or some other mechanism that throws `KafkaExtractEnd`, e.g. creating a file in a directory, or some other process signaling). 

For now, this is only used in tests, otherwise there's no way to "stop" the `KafkaExtract.run()` iterator.